src/query/functions-v2/src/scalars/string.rs

```diff
@@ -443,74 +433,94 @@ fn vectorize_string_to_string_2_arg(
-    registry.register_1_arg::<NumberType<u64>, StringType, _, _>(
+    registry.register_passthrough_nullable_1_arg::<NumberType<u64>, StringType, _, _>(
         "bin",
         FunctionProperty::default(),
         |_| None,
-        |val| format!("{:b}", val).as_bytes().to_vec(),
+        vectorize_with_writer_1_arg::<NumberType<u64>, StringType>(|val, writer| {
+            write_and_commit!(writer, "{:b}", val)
+        }),
     );
-    registry.register_1_arg::<NumberType<i64>, StringType, _, _>(
+    registry.register_passthrough_nullable_1_arg::<NumberType<i64>, StringType, _, _>(
         "bin",
         FunctionProperty::default(),
         |_| None,
-        |val| format!("{:b}", val).as_bytes().to_vec(),
+        vectorize_with_writer_1_arg::<NumberType<i64>, StringType>(|val, writer| {
+            write_and_commit!(writer, "{:b}", val)
+        }),
     );
-    registry.register_1_arg::<NumberType<f64>, StringType, _, _>(
+    registry.register_passthrough_nullable_1_arg::<NumberType<f64>, StringType, _, _>(
         "bin",
         FunctionProperty::default(),
         |_| None,
-        |val| {
-            let val = val as f64;
-            let ret = if val.ge(&0f64) {
-                format!(
-                    "{:b}",
-                    val.max(i64::MIN as f64).min(i64::MAX as f64).round() as i64
-                )
+        vectorize_with_writer_1_arg::<NumberType<f64>, StringType>(|val, writer| {
+            if (val as f64).ge(&0f64) {
+                let v = val.max(i64::MIN as f64).min(i64::MAX as f64).round() as i64;
+                write_and_commit!(writer, "{:b}", v)
             } else {
-                format!(
-                    "{:b}",
-                    val.max(u64::MIN as f64).min(u64::MAX as f64).round() as u64
-                )
-            };
-            ret.as_bytes().to_vec()
-        },
+                let v = val.max(u64::MIN as f64).min(u64::MAX as f64).round() as u64;
+                write_and_commit!(writer, "{:b}", v)
+            }
+        }),
     );
 }
 
 fn register_oct(registry: &mut FunctionRegistry) {
-    registry.register_1_arg::<NumberType<u64>, StringType, _, _>(
+    registry.register_passthrough_nullable_1_arg::<NumberType<u64>, StringType, _, _>(
         "oct",
         FunctionProperty::default(),
         |_| None,
-        |val| val.oct_string().as_bytes().to_vec(),
+        vectorize_with_writer_1_arg::<NumberType<u64>, StringType>(|val, writer| {
+            write_and_commit!(writer, "{}", oct_u64(val))
+        }),
     );
-    registry.register_1_arg::<NumberType<i64>, StringType, _, _>(
+    registry.register_passthrough_nullable_1_arg::<NumberType<i64>, StringType, _, _>(
         "oct",
         FunctionProperty::default(),
         |_| None,
-        |val| val.oct_string().as_bytes().to_vec(),
+        vectorize_with_writer_1_arg::<NumberType<i64>, StringType>(|val, writer| {
+            write_and_commit!(writer, "{}", oct_i64(val))
```

Comment:
Can we remove `oct_i64` as `bin` function did ? It creates temporary string memory.

Reply:
Sure.